modules/EnsEMBL/Web/Component/Gene/ComparaOrthologs.pm

```diff
@@ -0,0 +1,110 @@
+=head1 LICENSE
+
+Copyright [2009-2024] EMBL-European Bioinformatics Institute
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+     http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+
+=cut
+
+package EnsEMBL::Web::Component::Gene::ComparaOrthologs;
+
+use strict;
+use warnings;
+
+our $GENE_TREE_CONSTANTS = {
+  default => {
+    name => 'Metazoa',
+    url_part => 'Compara_Tree'
+  },
+  protostomes => {
+    name => 'Protostomes',
+    url_part => 'Protostomes_Tree'
+  },
+  insects => {
+    name => 'Insects',
+    url_part => 'Insects_Tree'
+  },
+  pangenome_drosophila => {
+    name => 'Drosophilidae',
+    url_part => 'Drosophilidae_Tree'
+  }
+};
+
+# Declare clusterset ids separately in an array, to keep their order
+our $METAZOA_CLUSTERSETS = ['default', 'protostomes', 'insects', 'pangenome_drosophila'];
+
+sub create_gene_tree_links {
+  my $self = shift;
+  my $params = shift;
+
+  my $hub          = $self->hub;
+  my $gene_availability = $params->{gene_availability};
+  my $stable_id = $params->{stable_id};
+  my $orthologue = $params->{orthologue};
+  my $cdb = $params->{cdb};
+
+  my $is_pan = $cdb =~/compara_pan_ensembl/;
+
+  my $current_gene_clusterset_ids = $self->get_current_gene_clusterset_ids($gene_availability);
+  my $common_clusterset_ids = [grep { $orthologue->{'homologue'}->has_GeneTree($_) } @$current_gene_clusterset_ids];
+
+  my $is_single_clusterset = scalar @{ $common_clusterset_ids } == 1;
+
+  my @link_str_parts;
+
+  foreach my $clusterset_id (@$common_clusterset_ids) {
+    my $gene_tree_constants = $self->get_gene_tree_constants_for_clusterset($clusterset_id);
+    my $gene_tree_name = $gene_tree_constants->{name};
+    my $link_text = $is_single_clusterset ? 'View Gene Tree' : "View $gene_tree_name Gene Tree";
+    my $url_part = $is_pan ? 'PanComparaTree' : $gene_tree_constants->{url_part};
```

Comment:
Hi @twalsh-ebi. Reading this code again got me wondering: should the `PanCompara` tree link generation be moved out of this for loop? According to the way this code is written, isn't PanCompara tree link a special case that does not depend on any clustersets that this loop is iterating over?

Reply:
Hi @azangru .. PanCompara is indeed a special case, though the Pan gene-tree link still works as it should because it has one clusterset_id — `default` — which happens to share a name with one of the `$METAZOA_CLUSTERSETS`.

AFAIK there are no plans to get rid of the `default` clusterset in either Metazoa or Pan Compara, so it should be pretty safe. But if you want to err on the side of caution, separating out the `PanCompara` tree link generation would certainly make sense. Maybe this method could call `$self->PREV::create_gene_tree_links` as soon as it knows it's dealing with a Pan Compara orthologue?